### ISSUE TITLE
Huge image size for a simple PHP App

### ISSUE THREAD

**Issue by arkkanoid** [deis/deis#4770]
Hi, 
I'm deploying a simple PHP App (it's the php-getting-started example of Heroku), and it runs perfectly. My doubt is when I check the image size, 1.516GB , wow it's really huge for this simple App. 
I deployed another PHP applications and I get exactly the same size.
I saw that the Application is builded from deis/slugrunner that has a size of 833MB (https://imagelayers.io/?images=deis%2Fslugrunner:latest).

I want to deploy a lot of apps, I will have at least this image size for all my apps? Could I reduce it?

**Comment by intellix** [comment]
Aye the cedar stack is huge, it pretty much installs everything despite you only wanting to run PHP: https://github.com/heroku/stack-images/blob/master/bin/cedar-14.sh

When I asked about this, I was recommended to push Docker images rather than rely on the php-buildpack from cedar.

**Comment by nathansamson** [comment]
My understanding is that almost all of this 1.5GB is shared over all deployed apps.
only a limited amount (100MB for my typical rails app which has rather a lot of assets), is used per deployment

**Comment by arkkanoid** [comment]
@nathansamson I've two similar PHP apps deployed in the same node and have two images of 1.516GB eachone.

REPOSITORY                         TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
10.21.2.109:5000/skiing-bookworm   v2                  ac13af058387        3 hours ago         1.516 GB
10.21.2.109:5000/trendy-seedcake   v2                  04c1e48efce1        3 days ago          1.516 GB

**Comment by Joshua-Anderson** [comment]
@arkkanoid in total the images are that large, but what @nathansamson is saying is that the cedar stack is a common layer shared between them, so it only exists on disk once.

**Comment by bacongobbler** [comment]
@arkkanoid if you're using the buildpack workflow, you'll get the entire [Heroku cedar stack](https://hub.docker.com/r/heroku/cedar) + whatever the buildpack installs onto the image. It's a very generic build image, so it's not going to be optimized for disk space. You can slim the image down yourself by [supplying your own Dockerfile](http://docs.deis.io/en/latest/using_deis/using-dockerfiles/), though.

**Comment by nathansamson** [comment]
@arkkanoid Note the virtual in "virtual size", the actual size on disk for these to images combined will probably be around 1.6-1.75GB (depending on your app). I don't know a method to actually check my theory for this particular case, which is a shame...

My understanding is that moving onto a Dockerfile workflow will actually be worse if you have multiple apps. Yes you can optimize each app and maybe get the size down to 500MB / app, but there will be less/nothing shared, so in the end you will end up in more storage in registry (which gets amplified by factor 3), and more diskspace on actual nodes when pulling the image to run, and resutling in slower pulls as there will be more network traffic...

If one of the maintainers can confirm my logic that would be nice, just reassuring to me I still have a decent understanding of how Deis / docker works...

**Comment by bacongobbler** [comment]
Warning, wall of text incoming...

@nathansamson I have to agree but also disagree with you here. I do agree with you that optimizing each app individually will net larger disk space utilization (unless the optimization is dramatic, see below on static Go binary workflow). What I disagree with is the actual size on disk you mentioned. Note that the original cedar stack size [is roughly 500MB in size](https://hub.docker.com/r/heroku/cedar/tags/) but the final image size is 1.5GB. That means that each build has about 1GB of unshared libraries between each build. The total disk space utilization for any given app will be 500MB + x * 1GB, where x is the number of releases that contain a Build. Typically this is x - 1 (v2 is typically the first release with a build). If @arkkanoid wanted to improve upon this, the best way would be to ensure that the apps and each apps releases share as much resources as possible.

To give a mathematical equation to this, let's assume one app with 3 builds. We'll make calculations for the total disk usage for this app with two different builds (Heroku vs Dockerfile). Let's assume the following:

 - the final image size will be 1.5GB for BOTH types
 - the Heroku build image is 512MB for the cedar stack, and the buildpack installs the other 1GB of dependencies
 - the Dockerfile image is 1280MB in size, then at the last minute statically compiles assets and copies configuration + the app into the image (256MB is a *very* generous assumption)

In both cases, the **total** disk space utilization will be (in megabytes)

Heroku: 512 + 3 * 1024 = 3584MB

Dockerfile: 1280 + 3 * 256 = 2048MB

To answer your original question, @nathansamson...

> the actual size on disk for these to images combined will probably be around 1.6-1.75GB

Using the above calculations, it'll actually be about 2.5GB of disk usage since each build only shares ~500MB. If they shared (nearly) everything, then the actual disk size would be closer to your assumption.

**TL;DR:** the best way to reduce disk usage is to ensure that your application builds share as many layers together as possible, and that the delta between each build is kept to a minimum.

As an aside, I'd also like to point out that workflows can come in different shapes and sizes, depending on the capabilities of the language. In particular, one workflow that has become semi-standard with Go apps is to compile the application [as a static binary](https://developer.atlassian.com/blog/2015/07/osx-static-golang-binaries-with-docker/), inject it into a Docker `scratch` image through their CI pipeline (resulting in a ~5MB image) then using `deis pull` to import the image into Deis. This is a dramatic improvement from 1.5GB or even [247MB](https://hub.docker.com/r/library/golang/tags/) for the 1.5.1 image, weighing in at about 2.5% of the latter's image size.

**Comment by krancour** [comment]
@arkkanoid does this information help you at all?  We should consider closing this since there isn't anything actionable here-- unless as @bacongobbler mentioned you feel docs should address image size.  That should probably be a separate issue.

**Comment by arkkanoid** [comment]
Sure, we can close this issue. Thanks all for your comments!
